go/app/main.go

```diff
@@ -20,22 +24,140 @@ type Response struct {
 	Message string `json:"message"`
 }
 
+/*
+list of item
+*/
+type ItemList struct {
+	Items []Item `json:"items"`
+}
+
+/*
+name, category and image of goods
+*/
+type Item struct {
+	Name     string `json:"name"`
+	Category string `json:"category"`
+	Image    string `json:"image_name"`
+}
+
+/*
+e.GET("/", root)
+*/
 func root(c echo.Context) error {
 	res := Response{Message: "Hello, world!"}
 	return c.JSON(http.StatusOK, res)
 }
 
+var itemlist ItemList
+
+/*
+e.POST("/items", addItem)
+*/
 func addItem(c echo.Context) error {
+	var item Item
+
 	// Get form data
-	name := c.FormValue("name")
-	c.Logger().Infof("Receive item: %s", name)
+	item.Name = c.FormValue("name")
+	item.Category = c.FormValue("category")
+	imagefile, err := c.FormFile("image")
+	if err != nil {
+		return err
+	}
+
+	//画像ファイルを開く
+	src, err := imagefile.Open()
+	if err != nil {
+		return err
+	}
+	defer src.Close()
 
-	message := fmt.Sprintf("item received: %s", name)
+	//hash化
+	h := sha256.New()
+	if _, err := io.Copy(h, src); err != nil { //srcからhへ中身をコピー
+		return err
+	}
+	str_hash_sha256 := fmt.Sprintf("%x", h.Sum(nil))
+	item.Image = str_hash_sha256 + ".jpg"
+
+	//imagesフォルダに画像ファイルを作成
+	dst, err := os.Create(fmt.Sprintf("images/%s", item.Image))
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+	if _, err = io.Copy(dst, src); err != nil { //srcからdstへ中身をコピー
+		return err
+	}
+
+	// add item to list
+	itemlist.Items = append(itemlist.Items, item)
+
+	//open file  if it doesn't exist, create file
+	jsonfile, err := os.OpenFile("items.json", os.O_WRONLY|os.O_CREATE, 0664)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer jsonfile.Close()
+
+	//encode
+	encoder := json.NewEncoder(jsonfile)
+	if err := encoder.Encode(itemlist); err != nil {
+		log.Fatal(err)
+	}
+
+	c.Logger().Infof("Receive item: %s, %s, %s", item.Name, item.Category, item.Image)
+	message := fmt.Sprintf("item received: %s", item.Name)
 	res := Response{Message: message}
 
 	return c.JSON(http.StatusOK, res)
 }
 
+func decodeJson() ItemList {
+	jsonfile, err := os.Open("items.json")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer jsonfile.Close()
+
+	//decode
+	var itemlist ItemList
+	decoder := json.NewDecoder(jsonfile)
+	if err := decoder.Decode(&itemlist); err != nil {
+		log.Fatal(err)
+	}
+
+	return itemlist
+}
+
+/*
+e.GET("/items". getItemList)
+*/
+func getItemList(c echo.Context) error {
+	itemlist := decodeJson()
+
+	return c.JSON(http.StatusOK, itemlist)
+}
+
+/*
+e.GET("/items/:id", getItemDetail)
+*/
+func getItemDetail(c echo.Context) error {
+	id, err := strconv.Atoi(c.Param("id")) //string to int
+	if err != nil {
+		return err
+	}
+	itemlist := decodeJson()
+
+	if id <= 0 || id > len(itemlist.Items) {
+		log.Fatal(err)
```

Comment:
[nits] log.Fatal を使わないで、できればerrorをreturnしてあげるようにしましょう。他のエラー処理も同様です。
エラーコードはその場所によって適切なコードを利用できると良いです。

```suggestion
     return c.JSON(http.StatusInternalServerError, Response{Message: err.Error()})

```

Reply:
[439c998](https://github.com/hinako1217/mercari-build-training/pull/6/commits/439c998e1bc1108a09b890b52bd66e5b49f979bf)
修正しました